Mathlib/Algebra/Group/Basic.lean

```diff
@@ -733,3 +733,28 @@ theorem div_eq_div_iff_div_eq_div : a / b = c / d ↔ a / c = b / d := by
   rw [div_eq_iff_eq_mul, div_mul_eq_mul_div, div_eq_iff_eq_mul', mul_div_assoc]
 
 end CommGroup
+
+/-- If a binary function from a type equipped with a total relation `r` to a monoid is
+  anti-symmetric (i.e. satisfies `f a b * f b a = 1`), in order to show it is multiplicative
+  (i.e. satisfies `f a c = f a b * f b c`), we may assume `r a b` and `r b c` are satisfied. -/
+@[to_additive additive_of_IsTotal "If a binary function from a type equipped with a total relation
+  `r` to an additive monoid is anti-symmetric (i.e. satisfies `f a b + f b a = 0`), in order to show
+  it is multiplicative (i.e. satisfies `f a c = f a b + f b c`), we may assume `r a b` and `r b c`
+  are satisfied."]
+lemma multiplicative_of_IsTotal [Monoid β] (f : α → α → β) (r : α → α → Prop) [t : IsTotal α r]
+    (hswap : ∀ a b, f a b * f b a = 1)
+    (hmul : ∀ {a b c}, r a b → r b c → f a c = f a b * f b c)
+    (a b c : α) : f a c = f a b * f b c := by
+  have h : ∀ b c, r b c → f a c = f a b * f b c := by
+    intros b c hbc
+    obtain hab | hba := t.total a b
+    · exact hmul hab hbc
+    obtain hac | hca := t.total a c
+    · rw [hmul hba hac, ← mul_assoc, hswap a b, one_mul]
+    · rw [← one_mul (f a c), ← hswap a b, hmul hbc hca, mul_assoc, mul_assoc, hswap c a, mul_one]
+  obtain hbc | hcb := t.total b c
+  · exact h b c hbc
+  · rw [h c b hcb, mul_assoc, hswap c b, mul_one]
```

Comment:
To reviewer:

I can't quite translate the [Lean 3 proof](https://github.com/leanprover-community/mathlib/pull/18080/files#diff-8626c6c73cd16bf773b579c3fa072a3cc334e223013d4cdc520288ac34403d33R662-R677) faithfully, so the order is a bit different from original. If you are able to make the `have := _` trick in Lean 3 work in Lean 4, please let me know! I'm also setting up mathport oneshot.

By the way, is there a counterpart to `#where` in Lean 4?